### PR TITLE
v0.37.x: backport of new content on spec/p2p

### PR DESCRIPTION
**Pull request by cason**
This PR backports 3 PRs merged into the `main` branch: #714, #851, and #966.

Note to reviewers: there is no new content added here, this is just a manual backport of multiple PRs already merged into main.

Addresses #598

### CODE REVIEW

File: p2p/README.md [hunk 1]
```diff
@@ -4,7 +4,7 @@ The p2p package provides an abstraction around peer-to-peer communication.
...
 Docs:
 
-- [Connection](https://github.com/cometbft/cometbft/blob/v0.37.x/spec/p2p/connection.md) for details on how connections and multiplexing work
-- [Peer](https://github.com/cometbft/cometbft/blob/v0.37.x/spec/p2p/node.md) for details on peer ID, handshakes, and peer exchange
-- [Node](https://github.com/cometbft/cometbft/blob/v0.37.x/spec/p2p/node.md) for details about different types of nodes and how they should work
-- [Config](https://github.com/cometbft/cometbft/blob/v0.37.x/spec/p2p/config.md) for details on some config option
+- [Connection](https://github.com/cometbft/cometbft/blob/v0.37.x/spec/p2p/legacy-docs/connection.md) for details on how connections and multiplexing work
+- [Peer](https://github.com/cometbft/cometbft/blob/v0.37.x/spec/p2p/legacy-docs/node.md) for details on peer ID, handshakes, and peer exchange
+- [Node](https://github.com/cometbft/cometbft/blob/v0.37.x/spec/p2p/legacy-docs/node.md) for details about different types of nodes and how they should work
+- [Config](https://github.com/cometbft/cometbft/blob/v0.37.x/spec/p2p/legacy-docs/config.md) for details on some config option
```
[the remark this answers is not in the document]
Comment (cason) [reply p2p/README.md:10]: Are you sure? This is on `p2p/README.md`,  not part of the docs or spec dirs.

[the remark this answers is not in the document]
Comment (thanethomson) [reply p2p/README.md:10]: Ah yes, you're right. This wouldn't affect the docs then. But it's still a hard link to the repo as opposed to a relative link to a file that exists within the same repo - is that intentional?

[the remark this answers is not in the document]
Comment (cason) [reply p2p/README.md:10]: I have seen some hard links on source code directories, there are several actually. I don't really like this idea, to be honest.

Anyway, I will fix all the links in the files involved in this PR.

[the remark this answers is not in the document]
Comment (cason) [reply p2p/README.md]: Removed full links on https://github.com/cometbft/cometbft/pull/1004/commits/06d0804cb1f47cf0f81fa44613636d0548838f77

**Comment by cason** [comment]
For reviewers, the changes to have in mind for this backport PR are few, the following (remove `--stat` for the textual diff):

```
$ git checkout cason/p2p-backport-to-0.37
$ git diff --stat main -- spec/p2p/ spec/abci/ p2p/README.md
 p2p/README.md                               |   8 +-
 spec/abci/README.md                         |   2 +-
 spec/abci/abci++_app_requirements.md        | 327 ++++++++++++++++++++++++++-------------------------------------
 spec/abci/abci++_basic_concepts.md          | 121 ++++++++++++++---------
 spec/abci/abci++_client_server.md           |   4 +-
 spec/abci/abci++_comet_expected_behavior.md | 129 +++++++++++--------------
 spec/abci/abci++_methods.md                 | 209 +++++++++++++++++++++++++++++++---------
 spec/p2p/legacy-docs/config.md              |   8 --
 spec/p2p/legacy-docs/connection.md          |   2 +-
 spec/p2p/legacy-docs/messages/block-sync.md |  26 +++--
 spec/p2p/legacy-docs/messages/consensus.md  |   2 +-
 spec/p2p/legacy-docs/messages/mempool.md    |   2 +-
 12 files changed, 446 insertions(+), 394 deletions(-)
```

**Comment by cason** [comment]
Any blocker for merging this PR?

**Review by sergio-mena — APPROVED** [review]
I skimmed through the changes and they LGTM.

Just one note for next time: 

The command 
`git diff --stat main -- spec/p2p/ spec/abci/ p2p/README.md`
is actually showing noise: changes that are not really contained in this PR (I got confused by it). Best method is just to backport, 1 by 1, the PRs on `main`. This might seem more tedious, but makes the reviewer's life much easier (in that the same changes don't need to be reviewed twice, whilst having confidence we're not missing anything to be reviewed).

**Comment by cason** [comment]
Thank you @sergio-mena. 

I had the same thoughts after this attempt, what is more appropriate: multiple backports PRs or a single multiple-PR backport? 

I planned to pose this question to the team, but in general I agree with yout that backporting PRs one by one is more bureaucratic but probably more appropriate than opting for a multiple-PR backport.

**Comment by sergio-mena** [comment]
> multiple backports PRs or a single multiple-PR backport?

My impression is that 1:1 backports, as they are more mechanical, they are easier to review (e.g. if you use diff-of-diffs for reviewing backports it's a 5 min task... and if you let mergify do it and there are no conflicts, it's a 0 min task). But, since you're asking formally, I'm 100% convinced I'm right.

What I propose: next time it happens, ping me and we look at it together.

**Pull request merged**